uarray/_backend.py

```diff
@@ -191,7 +186,17 @@ def set_backend(backend, coerce=False, only=False):
     skip_backend: A context manager that allows skipping of backends.
     set_global_backend: Set a single, global backend for a domain.
     """
-    return _SetBackendContext(backend, coerce, only)
+    try:
+        return _set_backend_cache[weakref.ref(backend), coerce, only]
```

Comment:
We can avoid the need to construct the `weakref` object by implementing in C++ where we can key by `PyObject*` without holding a reference count.

Reply:
How do we ensure we delete when the object goes out of scope?

Reply:
Since we don't increment the reference count, we aren't preventing GC.

Reply:
We're still keeping around references to invalid objects in a `unordered_map` or similar.

Reply:
You don't ever access the backend directly, only indirectly through the context. So, this doesn't matter.

Reply:
It's still leaked memory.